webservice/components/request_adapter.py

```diff
@@ -4,30 +4,51 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
 import requests
+from oauthlib.oauth2 import BackendApplicationClient
+from requests_oauthlib import OAuth2Session
 
 from odoo.addons.component.core import Component
 
 
+def build_session_oauth2(token_url, client_id, client_secret, audience=None):
+    """helper function get an authenticated OAuth2Session"""
+    client = BackendApplicationClient(client_id=client_id)
+    oauth = OAuth2Session(client=client)
+    # TODO: handle token renewal
+    oauth.fetch_token(
+        token_url=token_url,
+        client_id=client_id,
+        client_secret=client_secret,
+        audience=audience,
+    )
+    return oauth
+
+
 class BaseRestRequestsAdapter(Component):
     _name = "base.requests"
     _webservice_protocol = "http"
     _inherit = "base.webservice.adapter"
 
     # TODO: url and url_params could come from work_ctx
     def _request(self, method, url=None, url_params=None, **kwargs):
+        Session = self._get_session_class()
```

Comment:
Wouldn't be easier to not modify this adapter, you should create a new one?
The protocol can be `http_oauth`

Reply:
I rewrote the PR, I had come to the same conclusion. The result I feel is much cleaner.